grpc/server.go

```diff
@@ -24,55 +26,93 @@ var CodedError = status.Errorf
+		}
 	}
-	registerer(server, impl)
 
-	healthServer := health.NewServer()
-	healthpb.RegisterHealthServer(server, healthServer)
+	ret := &serverBuilder{cfg: c, services: s}
+	ret = ret.Add(&healthpb.Health_ServiceDesc, health.NewServer())
+	return ret
+}
 
-	start := func() error {
-		return cmd.FilterShutdownErrors(server.Serve(listener))
+// Add registers a new service (consisting of its description and its
+// implementation) to the set of services which will be exposed by this server.
+// It returns the modified-in-place serverBuilder so that calls can be chained.
+// If there is an error adding this service, it will be exposed when .Build() is
+// called.
+func (sb *serverBuilder) Add(desc *grpc.ServiceDesc, impl any) *serverBuilder {
+	s, ok := sb.services[desc.ServiceName]
+	if !ok {
+		// If this service doesn't have its own config stanza, instead initialize it
+		// with pieces from the server-level config.
+		s = service{cfg: cmd.GRPCServiceConfig{ClientNames: sb.cfg.ClientNames}}
 	}
-	stop := func() {
-		healthServer.Shutdown()
-		server.GracefulStop()
+
+	if s.desc != nil || s.impl != nil {
+		// We've already registered a service with this same name, error out.
+		sb.err = fmt.Errorf("attempted double-registration of gRPC service %q", desc.ServiceName)
+		return sb
 	}
-	return start, stop, nil
+
+	s.desc = desc
+	s.impl = impl
+	sb.services[desc.ServiceName] = s
+
+	return sb
 }
 
-// NewServer creates a gRPC server that uses the provided *tls.Config, and
-// verifies that clients present a certificate that (a) is signed by one of
-// the configured ClientCAs, and (b) contains at least one
-// subjectAlternativeName matching the accepted list from GRPCServerConfig.
-func NewServer(c *cmd.GRPCServerConfig, tlsConfig *tls.Config, statsRegistry prometheus.Registerer, clk clock.Clock, interceptors ...grpc.UnaryServerInterceptor) (*grpc.Server, net.Listener, error) {
+// Build creates a gRPC server that uses the provided *tls.Config and exposes
+// all of the services added to the builder. It also exposes a health check
+// service. It returns two functions, start() and stop(), which should be used
+// to start and gracefully stop the server.
+func (sb *serverBuilder) Build(tlsConfig *tls.Config, statsRegistry prometheus.Registerer, clk clock.Clock, interceptors ...grpc.UnaryServerInterceptor) (func() error, func(), error) {
+	if sb.err != nil {
+		return nil, nil, sb.err
+	}
+
+	for serviceName := range sb.cfg.Services {
+		_, ok := sb.services[serviceName]
+		if !ok {
+			return nil, nil, fmt.Errorf("gRPC service %q configured but not registered", serviceName)
+		}
+	}
+
 	if tlsConfig == nil {
 		return nil, nil, errNilTLS
 	}
+
 	acceptedSANs := make(map[string]struct{})
-	for _, name := range c.ClientNames {
+	for _, name := range sb.cfg.ClientNames {
 		acceptedSANs[name] = struct{}{}
 	}
+	for _, service := range sb.services {
+		for _, name := range service.cfg.ClientNames {
+			acceptedSANs[name] = struct{}{}
```

Comment:
This is a little tricksy so let's add a comment explaining it: The overall server has to accept connections from all clients that any one of its services accepts. Per-service restrictions will be imposed separately at request time.